rwgame/RWConfig.cpp

```diff
@@ -0,0 +1,575 @@
+    RWConfigLayer layer;
+
+#define RWARG(...)
+#define RWCONFIGARG(_RW_TYPE, _RW_NAME, _RW_DEFAULT, _RW_CONFPATH, _RW_CATEGORY, _RW_ARGMASK, _RW_ARGMETA, _RW_HELP)    \
+    layer._RW_NAME = _RW_DEFAULT;
+#define RWARG_OPT(...)
+#include "RWConfig.inc"
+#undef RWARG_OPT
+#undef RWCONFIGARG
+#undef RWARG
+
+    return layer;
+}
+
+static constexpr auto kConfigDirectoryName = "OpenRW";
+
+rwfs::path RWConfigParser::getDefaultConfigPath() {
+#if defined(RW_LINUX) || defined(RW_FREEBSD) || defined(RW_NETBSD) || \
+    defined(RW_OPENBSD)
+    char *config_home = getenv("XDG_CONFIG_HOME");
+    if (config_home != nullptr) {
+        return rwfs::path(config_home) / kConfigDirectoryName;
+    }
+    char *home = getenv("HOME");
+    if (home != nullptr) {
+        return rwfs::path(home) / ".config/" / kConfigDirectoryName;
+    }
+
+#elif defined(RW_OSX)
+    char *home = getenv("HOME");
+    if (home)
+        return rwfs::path(home) / "Library/Preferences/" / kConfigDirectoryName;
+
+#elif defined(RW_WINDOWS)
+    wchar_t *widePath;
+    auto res = SHGetKnownFolderPath(FOLDERID_RoamingAppData, KF_FLAG_DEFAULT,
+                                    nullptr, &widePath);
+    if (SUCCEEDED(res)) {
+        auto utf8Path = wideStringToACP(widePath);
+        return rwfs::path(utf8Path) / kConfigDirectoryName;
+    }
+#else
+    return rwfs::path();
+#endif
+
+    // Well now we're stuck.
+    RW_ERROR("No default config path found.");
+    return rwfs::path();
+}
+
+namespace {
+
+void flatten_ptree_recursive(std::map<std::string, std::string> &map, const std::string &subkey, const pt::ptree &ptree) {
+    for (const auto &[name, content] : ptree) {
+        auto key = subkey + "." + name;
+        map[key] = content.data();
+        flatten_ptree_recursive(map, subkey + "." + name, content);
+    }
+}
+std::map<std::string, std::string> flatten_ptree(const pt::ptree &ptree) {
+    std::map<std::string, std::string> result;
+    for (const auto &[name, content] : ptree) {
+        flatten_ptree_recursive(result, name, content);
+    }
+    return result;
+}
+
+std::string stripComments(const std::string &str) {
+    auto s = std::string(str, 0, str.find_first_of(";#"));
+    return s.erase(s.find_last_not_of(" \n\r\t") + 1);
+}
+
+template <typename T>
+struct Translator {
+};
+
+template <>
+struct Translator<std::string> {
+    using internal_type = std::string;
+    using external_type = std::string;
+    boost::optional<external_type> get_value(const internal_type &str) {
+        return stripComments(str);
+    }
+    boost::optional<internal_type> put_value(const external_type &str) {
+        return str;
+    }
+};
+
+template <>
+struct Translator<bool> {
+    using internal_type = std::string;
+    using external_type = bool;
+    boost::optional<external_type> get_value(const internal_type &str) {
+        boost::optional<external_type> res;
+        try {
+            res = std::stoi(stripComments(str)) != 0;
```

Comment:
You can use `lexical_cast` for it: https://github.com/rwengine/openrw/blob/master/rwcore/rw/casts.hpp

Reply:
I need to detect whether the conversion succeeded.

Reply:
Unless you know of a way to detect conversion failure, I prefer to keep it as it is.